core/unix/injector.c

```diff
@@ -1045,6 +1049,82 @@ our_ptrace_getvregs(pid_t pid, struct USER_VREGS_TYPE *vregs)
     struct iovec iovec = { vregs, sizeof(*vregs) };
     return our_ptrace(PTRACE_GETREGSET, pid, (void *)NT_FPREGSET, &iovec);
 }
+
+static long
+our_ptrace_get_sve(pid_t pid, struct user_sve_header *sve_header, size_t size)
+{
+    ASSERT(size >= sizeof(struct user_sve_header));
+    struct iovec iovec = { sve_header, size };
+    return our_ptrace(PTRACE_GETREGSET, pid, (void *)NT_ARM_SVE, &iovec);
+}
+
+static long
+our_ptrace_set_sve(pid_t pid, struct user_sve_header *sve_state)
+{
+    struct iovec iovec = { sve_state, sve_state->size };
+    return our_ptrace(PTRACE_SETREGSET, pid, (void *)NT_ARM_SVE, &iovec);
+}
+
+/* Get the SVE register state for the target process.
+ * Returns false if an error occurs.
+ * On success sve_state_out will be set to a heap allocated user_sve_header containing
+ * the SVE register state if the target process has SVE state active, or NULL if SVE
+ * state is not active.
+ * The caller is responsible for freeing the sve_state_out when it is no longer needed.
+ */
+static bool
+get_sve_state(pid_t pid, struct user_sve_header **sve_state_out)
+{
+    /* First we need to read the header to find out the vector length. */
+    struct user_sve_header sve_header = { 0 };
+    if (our_ptrace_get_sve(pid, &sve_header, sizeof(sve_header)) < 0) {
+        /* No SVE support. */
+        *sve_state_out = NULL;
+        return true;
+    }
+
+    if (sve_header.vl > MAX_SUPPORTED_SVE_VECTOR_LENGTH) {
+        /* DynamoRIO supports a maximum of 64-byte (512-bit) vectors so we can't
+         * attach to this process.
+         */
+        if (verbose) {
+            fprintf(stderr,
+                    "Unable to attach to process. The target process has an SVE vector "
+                    "length larger than the maxumum supported by DynamoRIO\n");
+        }
+        *sve_state_out = NULL;
+        return false;
+    }
+
+    /* Now we know the vector length we can allocate enough memory to contain the
+     * register data and call our_ptrace_get_sve() again to get the register data (if
+     * there is any).
+     */
+    const size_t full_sve_state_size =
+        SVE_PT_SIZE(sve_vq_from_vl(sve_header.vl), SVE_PT_REGS_SVE);
+    *sve_state_out = malloc(full_sve_state_size);
```

Comment:
Add a comment explaining that it is safe to use malloc because this is never run in DR managed mode. (I guess asserting !dr_app_running_under_dynamorio() would require linking on more libs than are currently pulled in?)